io/disk/lvsetup.py

```diff
@@ -74,37 +75,74 @@ def setUp(self):
 
         self.lv_snap_name = self.params.get(
             'lv_snapshot_name', default='avocado_sn')
-        self.ramdisk_basedir = self.params.get(
-            'ramdisk_basedir', default=os.path.join(self.workdir, 'ramdisk'))
-        self.ramdisk_sparse_filename = self.params.get(
-            'ramdisk_sparse_filename', default='virtual_hdd')
-
-        if 'delete' not in str(self.name.name):
-            if lv_utils.vg_check(self.vg_name):
-                self.cancel('Volume group %s already exists' % self.vg_name)
-            if lv_utils.lv_check(self.vg_name, self.lv_name):
-                self.cancel('Logical Volume %s already exists' % self.lv_name)
-            if lv_utils.lv_check(self.vg_name, self.lv_snap_name):
-                self.cancel('Snapshot %s already exists' % self.lv_snap_name)
+
+        if lv_utils.vg_check(self.vg_name):
+            self.cancel('Volume group %s already exists' % self.vg_name)
+        if lv_utils.lv_check(self.vg_name, self.lv_name):
+            self.cancel('Logical Volume %s already exists' % self.lv_name)
+        if lv_utils.lv_check(self.vg_name, self.lv_snap_name):
+            self.cancel('Snapshot %s already exists' % self.lv_snap_name)
 
         self.mount_loc = os.path.join(self.workdir, 'mountpoint')
         if not os.path.isdir(self.mount_loc):
             os.makedirs(self.mount_loc)
 
+        if self.lv_size:
+            # converting to megabytes
+            if self.lv_size.endswith('G'):
+                self.lv_size = int(self.lv_size.strip('G')) * 1024
+            elif self.lv_size.endswith('M'):
+                self.lv_size = int(self.lv_size.strip('M'))
+            else:
+                self.lv_size = int(self.lv_size) / 1024 / 1024
+
         if self.disk:
-            # converting bytes to megabytes, and using only 45% of the size
-            self.lv_size = int(lv_utils.get_diskspace(self.disk)) / 2330168
+            disk_size = lv_utils.get_devices_total_space(self.disk.split())
+            # converting bytes to megabytes
+            disk_size = disk_size / (1024 * 1024)
+
+            if self.lv_size:
+                if self.lv_size > disk_size:
+                    self.cancel("lv size provided more than size of disks")
+            else:
+                self.lv_size = disk_size
+
+            # Using only 45% of lv size, to accomodate lv snapshot also.
+            self.lv_size = (self.lv_size * 45) / 100
+            self.device = self.disk
         else:
-            self.lv_size = '1G'
-        self.lv_size = self.params.get('lv_size', default=self.lv_size)
+            if not self.lv_size:
+                self.lv_size = 1024
+
+            self.create_loop_dev()
+            self.device = self.loop
+            # Using only 45% of lv size, to accomodate lv snapshot also.
+            self.lv_size = (self.lv_size * 45) / 100
```

Comment:
This can be common for both blocks.

Reply:
Thanks @harish-24 . Have done that now, please check.